hub/core/meta/tensor_meta.py

```diff
@@ -78,16 +78,10 @@ def adapt(self, buffer: memoryview, shape: Tuple[int], dtype) -> memoryview:
         """
         dtype = np.dtype(dtype)
         if self.dtype and self.dtype != dtype.name:
-            if np.can_cast(dtype, self.dtype):
-                buffer = memoryview(
-                    np.cast[self.dtype](np.frombuffer(buffer, dtype=dtype)).tobytes()
-                )
-            else:
-                raise TensorDtypeMismatchError(
-                    self.dtype,
-                    dtype.name,
-                    self.htype,
-                )
+            buffer = memoryview(  # Already verified this is safe in tensor.extend
+                np.cast[self.dtype](np.frombuffer(buffer, dtype=dtype)).tobytes()
+            )
+
```

Comment:
btw, this is probably not an ideal way to handle re-casting. if the sample is compressed, this buffer will be decompressed and then recompressed

Reply:
agreed. let's shift all casting (both upcasting and downcasting) to data retrieval instead of upload. The check here can simply check dtype compatibility and while retrieving the np arrays, output samples can be cast to self.dtype if there's a mismatch.